### PR TITLE
add pipeline yml for syncing fabricbot.json aliases

### PR DESCRIPTION
**Pull request by greathongtu**
## Description

add pipeline yml file to sync fabricbot.json aliases, so that we can refer to this yml in devops in the future.

## Checklist

- [x] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** select appropriate branch. Cmdlets from Autorest.PowerShell should go to [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch. 
* **SHOULD** make the title of PR clear and informative, and in the present imperative tense. 
* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. Add changelog in description section if PR goes into [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** have approved design review for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) ([_Microsoft internal only_](../blob/main/CONTRIBUTING.md#onboarding)) with following situations
    * Create new module from scratch
    * Create new resource types which are not easy to conform to [Azure PowerShell Design Guidelines](../blob/main/documentation/development-docs/design-guidelines)
    * Create new resource type which name doesn't use module name as prefix
    * Have design question before implementation
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** introduce [breaking changes](../blob/main/documentation/breaking-changes/breaking-changes-definition.md) in Az minor release except preview version.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

File: .azure-pipelines/sync-aliases.yml [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  - task: PowerShell@2
+    displayName: Update json file
+    inputs:
+      targetType: inline
+      script: |
+        $username="v-hongtzhang"
+        $password='$(AZURE-PASS)'
+        $pair="{0}:{1}" -f ($username,$password)
+        $bytes = [System.Text.Encoding]::ASCII.GetBytes($pair)
+        $token = [System.Convert]::ToBase64String($bytes)
```
Comment (isra-fel) [inline .azure-pipelines/sync-aliases.yml:27]: This is a bad idea. This password is too important you shouldn't trust anyone to store it.

Wiki is also a repo, maybe you can clone the repo instead?

File: .azure-pipelines/sync-aliases.yml [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+- job: UpdateJson
+  displayName: Update Json
+  steps:
+  - checkout: self
+    persistCredentials: true
+  - task: PowerShell@2
+    displayName: Update json file
+    inputs:
+      targetType: inline
+      script: |
```
Comment (YanaXu) [inline .azure-pipelines/sync-aliases.yml:22]: The script is a little complicated. Can we save it in a ps1 file?

File: .azure-pipelines/sync-aliases.yml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  repositories:
+  - repository: self
+    type: git
+    ref: main
+
+trigger: none
+
+jobs:
+- job: UpdateJson
+  displayName: Update Json
```
Comment (isra-fel) [inline .azure-pipelines/sync-aliases.yml:12]:
```suggestion
  displayName: Update fabricbot.json
```

File: .azure-pipelines/sync-aliases.yml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        git checkout -b "Change_fabricbot.json_aliases"
+        
+        git add .
+        git commit -m "Use pipeline to sync fabricbot22.json"
+        
+        git remote set-url origin https://$(BotAccessToken)@github.com/Azure/azure-powershell.git;
+        git push origin Change_fabricbot.json_aliases --force
+  
+  - pwsh: |
+      $Title = "Change Fabricbot.json According To Wiki Page"
```
Comment (isra-fel) [inline .azure-pipelines/sync-aliases.yml:36]:
```suggestion
      $Title = "Sync Fabricbot.json According To ADO Wiki Page"
```
Fabricbot.json is 'F' uppercase or lowercase?

File: .azure-pipelines/sync-aliases.yml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+      $HeadBranch = "Change_fabricbot.json_aliases"
```
Comment (isra-fel) [inline .azure-pipelines/sync-aliases.yml:37]:
```suggestion
      $HeadBranch = "internal/sync-fabricbot-json"
```

File: .azure-pipelines/sync-aliases.yml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+      $BaseBranch = "main"
+      $Description = "Use powershell to sort taskType: scheduledAndTrigger part of fabricbot.json according to table of Service-Team-Label-and-Contact-List in wiki page"
```
Comment (isra-fel) [inline .azure-pipelines/sync-aliases.yml:39]:
```suggestion
      $Description = "This PR sync xxx from xxx."
```

**Review by isra-fel — CHANGES_REQUESTED** [review]
Suggestions of strings

File: .azure-pipelines/sync-aliases.yml [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+trigger: none
+
+jobs:
+- job: UpdateJson
+  displayName: Update fabricbot.json
+  steps:
+  - checkout: self
+  - pwsh: |
+      ./tools/Github/ParseWiki2Json.ps1 -AzureToken $(AzureToken)
```
Comment (wyunchi-ms) [inline .azure-pipelines/sync-aliases.yml:16]: `ADOToken` may be a better name.

File: tools/Github/ParseWiki2Json.ps1 [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+$pair="{0}:{1}" -f ($username,$password)
+$bytes = [System.Text.Encoding]::ASCII.GetBytes($pair)
+$token = [System.Convert]::ToBase64String($bytes)
+$headers = @{
+    Authorization = "Basic {0}" -f ($token)
+}
+
+$response = Invoke-RestMethod 'https://dev.azure.com/azure-sdk/internal/_apis/wiki/wikis/internal.wiki/pages?api-version=7.1-preview.1&path=/Engineering%20System/GitHub%20Repos/Issue%20Management/Service%20Team%20Label%20and%20Contact%20List&includeContent=True' -Headers $headers
+
+$rows = ($response.content -split "\n") | Select-Object -Skip 6
```
Comment (wyunchi-ms) [inline tools/Github/ParseWiki2Json.ps1:37]: Not a good idea to rely on the magic number. You can try use `ConvertFrom-Markdown` to parse this document.

Comment (xtR0d666) [reply tools/Github/ParseWiki2Json.ps1]: I have changed it to be: `$rows = ($response.content -split "\n") | Where-Object { $_ -like '|*'} | Select-Object -Skip 2`, which is more robust for potential changes. The number 2 here is to skip the table header and separator.

File: .azure-pipelines/sync-aliases.yml [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  - pwsh: |
+      ./tools/Github/ParseWiki2Json.ps1 -ADOToken $(ADOToken)
+    displayName: Update fabricbot.json file locally
+      
+  - task: PowerShell@2
+    displayName: Git commit and push
+    inputs:
+      targetType: inline
+      script: |
+        git pull origin main
```
Comment (wyunchi-ms) [inline .azure-pipelines/sync-aliases.yml:24]: Why need to pull origin main? It's finished in step `checkout`

File: .azure-pipelines/sync-aliases.yml [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    type: git
+    ref: main
+
+trigger: none
+
+jobs:
+- job: UpdateJson
+  displayName: Update fabricbot.json
+  steps:
+  - checkout: self
```
Comment (YanaXu) [inline .azure-pipelines/sync-aliases.yml:14]: remove it if it's not necessary.

File: .azure-pipelines/sync-aliases.yml [hunk 11]
```diff
@@ -0,0 +1,39 @@
+# Variable 'ADOToken' and 'BotAccessToken' was defined in the Variables tab
+schedules:
+- cron: "50 15 * * *"
+  displayName: 11:50 PM (UTC + 8:00) China Daily Build
+  branches:
+    include:
```
Comment (YanaXu) [inline .azure-pipelines/sync-aliases.yml:6]: remove "schedules".

File: tools/Github/CreatePR.ps1 [hunk 12]
```diff
@@ -64,4 +64,11 @@ $Description
 * **SHOULD NOT** adjust version of module manually in pull request
 "@
 $RequestBody = @{"title" = $Title; "body" = $PrBody; "head" = $HeadBranch; "base" = $BaseBranch }
-Invoke-WebRequest -Uri https://api.github.com/repos/Azure/azure-powershell/pulls -method POST -Headers $Headers -Body ($RequestBody | ConvertTo-Json)
\ No newline at end of file
+$Uri = "https://api.github.com/repos/Azure/azure-powershell/pulls"
+$PullRequests = Invoke-WebRequest -Uri $Uri -Headers $Headers | ConvertFrom-Json
+$ExistingPr = $PullRequests | Where-Object { $_.title -eq $Title }
```
Comment (wyunchi-ms) [inline tools/Github/CreatePR.ps1:69]: I mean add a switch parameter such as `FailOnPRExisted` and set the default value `False`. If other script care about the existing of PR, they can get the error. And script will be success for those who doesn't care about.

File: .azure-pipelines/sync-aliases.yml [hunk 13]
```diff
@@ -0,0 +1,39 @@
+# Variable 'ADOToken' and 'BotAccessToken' was defined in the Variables tab
+schedules:
+- cron: "50 15 * * *"
+  displayName: 11:50 PM (UTC + 8:00) China Daily Build
```
Comment (wyunchi-ms) [inline .azure-pipelines/sync-aliases.yml:4]: Please change the name, not Daily Build.

File: tools/Github/CreatePR.ps1 [hunk 14]
```diff
@@ -64,4 +66,17 @@ $Description
 * **SHOULD NOT** adjust version of module manually in pull request
 "@
 $RequestBody = @{"title" = $Title; "body" = $PrBody; "head" = $HeadBranch; "base" = $BaseBranch }
-Invoke-WebRequest -Uri https://api.github.com/repos/Azure/azure-powershell/pulls -method POST -Headers $Headers -Body ($RequestBody | ConvertTo-Json)
\ No newline at end of file
+
+$Uri = "https://api.github.com/repos/Azure/azure-powershell/pulls"
+$PullRequests = Invoke-WebRequest -Uri $Uri -Headers $Headers | ConvertFrom-Json
+$ExistingPr = $PullRequests | Where-Object { $_.title -eq $Title }
```
Comment (wyunchi-ms) [inline tools/Github/CreatePR.ps1:72]: I think no need to check the title, you just need to check whether there is a PR whose host is $Host and $base is $Base

File: tools/Github/CreatePR.ps1 [hunk 15]
```diff
@@ -64,4 +66,17 @@ $Description
...
+
+if ($ExistingPr -and $FailOnPRExisted) {
+    Write-Error "The PR named $Title already exists."
+    exit 1
+}
+if ($ExistingPr) {
+    Write-Host "The PR named $Title already exists. Skipping creation because FailOnPRExisted is set to $FailOnPRExisted."
+    exit 0
+}
```
Comment (wyunchi-ms) [inline tools/Github/CreatePR.ps1:81]: You can merge these two condition cases.

File: tools/Github/CreatePR.ps1 [hunk 16]
```diff
@@ -63,5 +65,24 @@ $Description
...
 * **SHOULD NOT** adjust version of module manually in pull request
 "@
+
 $RequestBody = @{"title" = $Title; "body" = $PrBody; "head" = $HeadBranch; "base" = $BaseBranch }
-Invoke-WebRequest -Uri https://api.github.com/repos/Azure/azure-powershell/pulls -method POST -Headers $Headers -Body ($RequestBody | ConvertTo-Json)
\ No newline at end of file
+$Uri = "https://api.github.com/repos/Azure/azure-powershell/pulls"
+
+$PullRequests = Invoke-WebRequest -Uri $Uri -Headers $Headers | ConvertFrom-Json
+$ExistingPr = $PullRequests | Where-Object { $_.head.ref -eq $HeadBranch -and $_.base.ref -eq $BaseBranch }
```
Comment (wyunchi-ms) [inline tools/Github/CreatePR.ps1:73]: I mean pass these two parameters as filter to github server. Please follow this doc I sent to you: https://docs.github.com/en/rest/pulls/pulls?apiVersion=2022-11-28#list-pull-requests

File: tools/Github/CreatePR.ps1 [hunk 17]
```diff
@@ -63,5 +65,23 @@ $Description
...
 * **SHOULD NOT** adjust version of module manually in pull request
 "@
+
 $RequestBody = @{"title" = $Title; "body" = $PrBody; "head" = $HeadBranch; "base" = $BaseBranch }
-Invoke-WebRequest -Uri https://api.github.com/repos/Azure/azure-powershell/pulls -method POST -Headers $Headers -Body ($RequestBody | ConvertTo-Json)
\ No newline at end of file
+$Uri = "https://api.github.com/repos/Azure/azure-powershell/pulls"
+
+$PrHead = @{"Accept" = "application/vnd.github+json"; "head" = $HeadBranch; "base" = $BaseBranch }
+$PullRequests = Invoke-RestMethod -Uri $Uri -Method GET -Headers $PrHead
```
Comment (wyunchi-ms) [inline tools/Github/CreatePR.ps1:73]: The response is an array, you need to check the length of the array.

**Review by wyunchi-ms — APPROVED** [review]

**Pull request merged**